src/ae_kqueue.c

```diff
@@ -101,62 +108,102 @@ static void aeApiFree(aeEventLoop *eventLoop) {
+                return -1;
+            state->nch = 0; /* rewind the changelist. */
+        }
     }
     return 0;
 }
 
 static void aeApiDelEvent(aeEventLoop *eventLoop, int fd, int mask) {
     aeApiState *state = eventLoop->apidata;
-    struct kevent ke;
-
-    if (mask & AE_READABLE) {
-        EV_SET(&ke, fd, EVFILT_READ, EV_DELETE, 0, 0, NULL);
-        kevent(state->kqfd, &ke, 1, NULL, 0, NULL);
+    /* At this point, we somehow receive deletion requests for events that
+     * are not registered in kqueue yet, which could causes kevent(2) to
+     * fail and return ENOENT. Therefore, we need to use aeEventLoop->events
+     * to mask out the events that are not registered in kqueue and get the
+     * valid events which are requested to be deleted. */
+    int delmask = eventLoop->events[fd].mask & mask;
+
+    /* Instead of applying events to kqueue one by one, we buffer events
+     * and apply them at once. */
+    if (delmask & AE_READABLE) {
+        EV_SET(state->changes + state->nch, fd, EVFILT_READ, EV_DELETE, 0, 0, NULL);
+        /* The current changelist is full, apply it to kqueue now and
+         * then rewind it to make rooms for follow-up events. */
+        if (++state->nch == EVENT_ARRAY_SIZE(state->changes)) {
+            if (kevent(state->kqfd, state->changes, state->nch, NULL, 0, NULL))
+                panic("aeApiDelEvent: kevent, %s", strerror(errno));
+            state->nch = 0; /* rewind the changelist. */
+        }
     }
-    if (mask & AE_WRITABLE) {
-        EV_SET(&ke, fd, EVFILT_WRITE, EV_DELETE, 0, 0, NULL);
-        kevent(state->kqfd, &ke, 1, NULL, 0, NULL);
+    if (delmask & AE_WRITABLE) {
+        EV_SET(state->changes + state->nch, fd, EVFILT_WRITE, EV_DELETE, 0, 0, NULL);
+        /* The current changelist is full, apply it to kqueue now and
+         * then rewind it to make rooms for follow-up events. */
+        if (++state->nch == EVENT_ARRAY_SIZE(state->changes)) {
+            if (kevent(state->kqfd, state->changes, state->nch, NULL, 0, NULL))
+                panic("aeApiDelEvent: kevent, %s", strerror(errno));
+            state->nch = 0; /* rewind the changelist. */
+        }
+    }
+
+    /* When it comes to EV_DELETE events, we don't defer but apply them immediately
+     * because the caller often closed the file descriptor right after they called
+     * aeDeleteFileEvent(), and kevent(2) would report ENOENT or EBADF if the changelist
+     * contained any closed file descriptors. */
+    if (state->nch > 0) {
+        if (kevent(state->kqfd, state->changes, state->nch, NULL, 0, NULL))
+            panic("aeApiDelEvent: kevent, %s", strerror(errno));
+        state->nch = 0; /* rewind the changelist. */
     }
 }
 
 static int aeApiPoll(aeEventLoop *eventLoop, struct timeval *tvp) {
     aeApiState *state = eventLoop->apidata;
     int retval, numevents = 0;
 
+    struct timespec ts, *timeout = NULL;
+    (void)(ts);
```

Comment:
What is this void for? I assume it's for an unused variable, but ts is used.

Reply:
You're right. Fixed.